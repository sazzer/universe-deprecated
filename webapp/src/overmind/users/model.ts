/** The state of the user record */
export type UserState = "LOADING" | "SAVING" | "UNKNOWN" | "LOADED";

/** The shape of a single user */
export interface User {
  /** The unqiue ID of the user */
  id: string;
  /** The unique username of the user */
  username: string;
  /** The display name of the user */
  displayName: string;
  /** The email address of the user, if we know what it is */
  email: string | null;
}

/** Representation of a user as known by the state */
export interface StoredUser {
  /** The user record, if we've currently got one */
  user: User | undefined;

  /** The current state of the user */
  state: UserState;
}

/** Error when loading a user that doesn't exist */
export class UnknownUserError extends Error {}
