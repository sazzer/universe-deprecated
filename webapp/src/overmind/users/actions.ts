import { Action } from "overmind";
import { User } from "./model";

/**
 * Overmind action for storing the details of a user
 */
export const storeUser: Action<User> = ({ state }, details: User) => {
  state.users.users[details.id] = {
    id: details.id,
    username: details.username,
    displayName: details.displayName,
    email: details.email
  };
  delete state.users.userStates[details.id];
};

export const fetchUser: Action<string, Promise<void>> = async (
  { state, actions, effects },
  userId: string
) => {
  state.users.userStates[userId] = "LOADING";
  const user = await effects.users.api.loadUser(userId);
  actions.users.storeUser(user);
};
