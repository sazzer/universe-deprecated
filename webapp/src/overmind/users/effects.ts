import { request, ProblemResponse } from "../../api";
import { User, UnknownUserError } from "./model";

/**
 * Retrieve a user from the API
 * @param id The ID of the user to retrieve
 * @return The details of the user
 */
async function loadUser(id: string): Promise<User> {
  try {
    const user = await request<User>({
      url: "/users/{id}",
      urlParams: {
        id
      },
      method: "GET"
    });

    return user.data;
  } catch (e) {
    throw new UnknownUserError();
  }
}

/** The API for working with users*/
export const api = {
  loadUser
};
