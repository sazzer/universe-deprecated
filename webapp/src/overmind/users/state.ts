export interface User {
  userId: string;
  username: string;
  displayName: string;
  email: string | null;
}

type State = {
  users: User[];
};

export const state: State = {
  users: []
};
