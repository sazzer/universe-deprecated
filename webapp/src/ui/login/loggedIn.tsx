import React from "react";
import { useOvermind } from "../../overmind";

/**
 * React Component to render the children only when there is an active logged-in session
 */
export const LoggedIn: React.FC = ({ children }) => {
  const { state } = useOvermind();

  if (
    state.authentication.isLoggedIn ||
    process.env.NODE_ENV === "development"
  ) {
    return <>{children}</>;
  } else {
    return <div>You must be logged in for this</div>;
  }
};
