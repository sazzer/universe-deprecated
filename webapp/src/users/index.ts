export * from "./model";
export * from "./userContext";
export * from "./api";
