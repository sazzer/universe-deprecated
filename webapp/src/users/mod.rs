mod model;
mod service;

pub use model::*;
