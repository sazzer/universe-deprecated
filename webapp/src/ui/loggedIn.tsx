import React from "react";
import { useTranslation } from "react-i18next";
import { useUser } from "../users";

/**
 * Wrapper around other components to only render if there is a currently logged in user
 */
export const LoggedIn: React.FC = ({ children }) => {
  const { hasUser } = useUser();
  const { t } = useTranslation();

  const body = hasUser ? children : t("errors.unauthenticated");

  return <>{body}</>;
};
