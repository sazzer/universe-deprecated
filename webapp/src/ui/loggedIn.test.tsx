import { UserContext, UserContextState } from "../users";

import { LoggedIn } from "./loggedIn";
import React from "react";
import { render } from "@testing-library/react";

describe("loggedIn", () => {
  test("When not logged in", () => {
    const userState: UserContextState = {
      hasUser: false,
      user: null,
      clearUser: () => {},
      storeUser: () => {}
    };

    const { container } = render(
      <UserContext.Provider value={userState}>
        <LoggedIn>Contents</LoggedIn>
      </UserContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test("When logged in", () => {
    const userState: UserContextState = {
      hasUser: true,
      user: null,
      clearUser: () => {},
      storeUser: () => {}
    };

    const { container } = render(
      <UserContext.Provider value={userState}>
        <LoggedIn>Contents</LoggedIn>
      </UserContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
