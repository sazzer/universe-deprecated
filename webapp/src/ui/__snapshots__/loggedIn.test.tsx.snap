// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loggedIn When logged in 1`] = `
<div>
  Contents
</div>
`;

exports[`loggedIn When not logged in 1`] = `
<div>
  You must be logged in to access this page
</div>
`;
